### DISCUSSION TITLE
Enable Async SelectAtom for SSR

### DISCUSSION THREAD

**Discussion by xd-hearst** [pmndrs/jotai#2033] [Bug report] [answered]
First of all, thanks for the great library. It has been tremendous help.

Second of all, Apologize if the following question is premature and shows a fundamental lack of knowledge. 

So ...

Scenario: we have a next js app, the app will have some initial values hydrated. However, the server side value might not be complete and we need to request new data if that is case. We only need request it once on the client to fill in the missing data.

An ideal solution would be we use `selectAtom` to get the value we want based on the atom and a key, if there is no value for a given key, we would like to trigger a fetch request on client side and update the atom. Otherwise, we have to prefetch the data use `useEffect` regardless if the value is ever needed. 

I have tried to get it to work. However, while an client side `async` selectAtom in conjunction with `loadable` works fine on the client side, it would never resolve if the atom is hydrated on the server side.

What would be the best way to handle this?

A server side hydration using `selectAtom` sandbox (code forked and modified from the next js example):

https://codesandbox.io/p/sandbox/hydration-next-js-client-async-update-qzxz2n

A client side `selectAtom` sandbox (code forked from another sample sandbox):

https://codesandbox.io/s/client-side-async-set-l4ljxm

Thank you very much

**Comment by xd-hearst** [comment]
I eventually use a synchronous selectAtom to retrieval the corresponding atom value. Then use a hook to check to see if the value is missing, then did a atom set to do the fetch and reset. Still would like to know if it is possible to use an async selectAtom in SSR. Thanks

**Reply by dai-shi** [reply]
I haven't looked at the code yet, but you wouldn't need `selectAtom` for this case.

My rough idea would be something like this:
```js
const fetchData = async () => ...
const prefetchedAtom = atom(null)
const dataAtom = atom((get) => get(prefetchedAtom) || fetchData())

const Component = ({ prefetchedData }) => {
  useHydrateAtoms([[prefetchedAtom, prefetcheData]])
  ...
}
```

**Reply by xd-hearst** [reply]
Thank you so much for the quick response.

Sorry for not making my case clear. 

The reason we have this server / client side dichotomy is we would like to keep the data from server side as lean as possible. Let's say I have a large data object with thousands of keys.

`const largeData = {testA: {...}, testB: {...}, ...}
`
```
const dataAtom = atom(null)

const Component = ({  }) => {
   const trimmedData = {testA: {...}}
  useHydrateAtoms([[dataAtom, trimmedData]])
  ...
}

```  

Component A needs to be rendered on the server/client side, it needs `data.testA`

Component B only needs to be rendered on the client side, it needs `data.testB`. At this point we want to do a fetch to make `data.testB` and the rest of data become available on the client side.

We would like to have componentA and componentB use the exact same hook to get the data by passing in the key, to accomplish this we would like to use `selectAtom`

Can you let me know if it makes sense?

thanks a lot 🙇

**Reply by dai-shi** [reply]
Apart from sever/client difference, what I meant by that `selectAtom` is not necessary is the following.

> We would like to have componentA and componentB use the exact same hook to get the data by passing in the key, to accomplish this we would like to use `selectAtom`

```js
const useSliceOfData = (sliceKey) => {
  const sliceAtom = useMemo(() => atom((get) => get(dataAtom)[sliceKey]), [sliceKey])
  const slice = useAtomValue(sliceAtom)
  return slice
}
```

**Reply by xd-hearst** [reply]
thank you. Oh, the `get` here looks pretty much the same as `selectAtom`.  I can change it. 

Wonder when it is better to use a `selectAtom`?

**Reply by dai-shi** [reply, marked as answer]
`selectAtom` (whose name in counter-intuitive) can be used for escape hatches. It's useful to violate declarative model with `equalityFn` and `prevSlice`.
